README.md
<div align="center">
	<h1>Roads</h1>
</div>
<div align="center">
	<a href="https://github.com/LastTalon/Roads/actions">
		<img src="https://github.com/LastTalon/Roads/workflows/CI/badge.svg" alt="CI Status">
	</a>
	<a href="https://coveralls.io/github/LastTalon/Roads?branch=master">
		<img src="https://coveralls.io/repos/github/LastTalon/Roads/badge.svg?branch=master" alt="Coverage Status">
	</a>
</div>
<br>

**Roads** is is a free software *Roblox* concurrency library.

## Usage
Roads uses [Rojo](https://rojo.space/) as a bridge between the filesystem and
Roblox Studio. If you want to contribute or use Roads for development, install
both to get started. The project contains a `default.project.json` that can be
used by Rojo to generate or sync with a Roblox save file.

## Documentation
The only documentation is currently in the source. This will be changing soon.

## Contributing
Contributions are welcome, please make a pull request!

Be sure to set up [Rojo](https://rojo.space/) to get started. Check out our
[contribution guide](CONTRIBUTING.md) for further information.

Please read our [code of conduct](CODE_OF_CONDUCT.md) when getting involved.

## License
Roads is free software available under the MIT license. See the
[license](LICENSE.md) for details.


CHANGELOG.md
# Roads Changelog
This file logs all notable changes for this project.

This project uses [semantic versioning](https://semver.org/).

## Unreleased Changes
* Added basic project info in `README.md`.
* Added MIT license in `LICENSE.md`.
* Added contributing guidelines in `CONTRIBUTING.md`.
* Added code of conduct in `CODE_OF_CONDUCT.md`.
* Added status badges for upcoming checks.
* Added issue templates.
	* Bug report
	* Feature request
* Added pull request template.
* Added Rojo project.
* Added CI testing and coverage reporting.


tests/Wait.spec.lua
--- Tests for the Wait function
--
-- @version 0.1.0, 2020-11-04
-- @since 0.1

return function()
	local module = game:GetService("ReplicatedStorage"):WaitForChild("Roads")
	local Wait = require(module:WaitForChild("Wait"))

	describe("Wait", function()
		it("should be a function", function()
			expect(Wait).to.be.a("function")
		end)

		it("should yield the current thread", function()
			local finished = false
			coroutine.wrap(function()
				Wait()
				finished = true
			end)()
			expect(finished).to.equal(false)
			Wait()
			expect(finished).to.equal(true)
		end)

		it("should yield for the duration passed", function()
			local time = tick()
			Wait(1)
			expect(tick()-time).to.be.near(1, 0.067)
		end)

		it("should return the actual duration yielded", function()
			local time = tick()
			local actual = Wait(1)
			expect(tick()-time).to.be.near(actual)
		end)
	end)
end


tests/Async.spec.lua
--- Tests for the Async function
--
-- @version 0.1.0, 2020-11-04
-- @since 0.1

return function()
	local module = game:GetService("ReplicatedStorage"):WaitForChild("Roads")
	local Async = require(module:WaitForChild("Async"))
	local Await = require(module:WaitForChild("Await"))

	describe("Async", function()
		it("should be a function", function()
			expect(Async).to.be.a("function")
		end)

		it("should only accept a function", function()
			expect(function() Async() end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got nil")
			expect(function() Async(false) end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got boolean")
			expect(function() Async(0) end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got number")
			expect(function() Async("") end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got string")
			expect(function() Async(Instance.new("Folder")) end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got userdata")
			expect(function() Async(
				coroutine.create(function() end))
			end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got thread")
			expect(function() Async({}) end).to.throw(
				"invalid arugment #1 to 'Async' (function expected, got table")
			expect(function() Async(function() end) end).never.to.throw()
		end)

		it("should return a function", function()
			local value = Async(function() end)
			expect(value).to.be.a("function")
		end)

		it("should not yield the calling thread", function()
			local activated = false
			local task = Async(function() activated = true end)()
			expect(activated).to.equal(false)
			Await(task)
			expect(activated).to.equal(true)
		end)

		it("should pass parameters to the async function when calling", function()
			local activated = false
			Await(Async(function(value) activated = value end)(true))
			expect(activated).to.equal(true)
		end)
	end)
end


lib/Task.lua
local extractStatus = require(script.Parent.ExtractStatus)
local taskManager = require(script.Parent.TaskManager)

local Task = {}
Task.__index = Task

Task.TaskStatus = require(script.Parent.StatusEnum)
Task.sleep = require(script.Parent.Sleep)

function Task.new(fn)
	local self = setmetatable({}, Task)
	if type(fn) == "function" then
		self.Coroutine = coroutine.create(fn)
	end
	self.Status = Task.TaskStatus.Pending
	self.Persist = false

	taskManager.register(self)
	return self
end

function Task:Completed()
	return self.Status == Task.TaskStatus.Completed
end

function Task:Canceled()
	return self.Status == Task.TaskStatus.Canceled
end

function Task:Start(...)
	if self.Status == Task.TaskStatus.Pending and coroutine.status(self.Coroutine) == "suspended" then
		self.Status, self.Value = extractStatus({coroutine.resume(self.Coroutine, ...)}, coroutine.status(self.Coroutine))
	end
	if not self.Persist and self.Status ~= Task.TaskStatus.Pending then
		local co = coroutine.create(function()
			Task.sleep()
			self:Destroy()
		end)
		coroutine.resume(co)
	end
end

function Task:Wait()
	while self.Status == Task.TaskStatus.Pending do
		Task.sleep()
		self:Start()
	end
end

function Task:Then(fulfill, reject)
	local fulfillTask
	local rejectTask

	if fulfill ~= nil and type(fulfill) == "function" then
		fulfillTask = Task.new(fulfill)
	end

	if reject ~= nil and type(reject) == "function" then
		rejectTask = Task.new(reject)
	end

	local deferred = Task.new(function()
		self:Wait()
		if self.Status == Task.TaskStatus.Completed then
			if fulfillTask ~= nil then
				fulfillTask:Start(unpack(self.Value))
			end
		else
			if rejectTask ~= nil then
				rejectTask:Start(self.Value)
			end
		end
	end)
	deferred:Start()

	return fulfillTask, rejectTask
end

function Task:Persist(persist)
	persist = persist or true
	if type(persist) == "boolean" then
		self.Persist = persist
	end
end

function Task:Destroy()
	self.Coroutine = nil
	self.Status = nil
	self.Value = nil
	taskManager.deregiser(self)
end

return Task


lib/Asynchronous.lua
local asynchronous = {}
local Task = require(script.Parent.Task)

asynchronous._version = "0.2.2"
asynchronous.TaskStatus = require(script.Parent.StatusEnum)
asynchronous.sleep = Task.sleep
asynchronous.yield = coroutine.yield

local function argTypeOrDie(arg, datatype, functionName, argPosition)
	local t = type(arg)
	if t ~= datatype then
		error("bad argument #" .. argPosition .. " to '" .. functionName .. "' (" ..datatype .. " expected, got " .. t .. ")", 3)
	end
end

function asynchronous.task(fn)
	argTypeOrDie(fn, "function", "task", 1)

	return Task.new(fn)
end

function asynchronous.async(fn)
	argTypeOrDie(fn, "function", "async", 1)

	local asyncFn = function(...)
		local task = asynchronous.task(fn)
		task:Start(...)
		return task
	end

	return asyncFn
end

function asynchronous.await(task)
	argTypeOrDie(task, "table", "await", 1)

	if task.Status == asynchronous.TaskStatus.Pending then
		task:Wait()
	end

	if task.Status == asynchronous.TaskStatus.Completed then
		return unpack(task.Value)
	else
		error(task.Value, 2)
	end
end

function asynchronous.callback(fn)
	argTypeOrDie(fn, "function", "callback", 1)

	local callbackFn = function(...)
		local task = asynchronous.task(fn)
		local args = {...}
		local fulfill
		local reject
		if type(args[#args]) == "function" then
			fulfill = args[#args]
			table.remove(args)
			if type(args[#args]) == "function" then
				reject = fulfill
				fulfill = args[#args]
				table.remove(args)
			end
		end

		task:Start(unpack(args))
		task:Then(fulfill, reject)

		return task
	end

	return callbackFn
end

return asynchronous


lib/init.lua
--- The Roads concurrency library
-- Provides access to all library classes and features.
--
-- @module Roads
-- @version 0.1.0, 2020-11-04
-- @since 0.1

local Task = require(script:WaitForChild("Task"))
local Scheduler = require(script:WaitForChild("Scheduler"))
local Wait = require(script:WaitForChild("Wait"))
local Delay = require(script:WaitForChild("Delay"))
local Async = require(script:WaitForChild("Async"))
local Await = require(script:WaitForChild("Await"))
local Callback = require(script:WaitForChild("Callback"))

local Roads = {}
local instance

Roads.__index = Roads

-- Library functions
Roads.Wait = Wait
Roads.Delay = Delay
Roads.Async = Async
Roads.Await = Await
Roads.Callback = Callback

function Roads.new()
	if instance == nil then
		local self = setmetatable({}, Roads)

		-- Library objects
		self.Task = Task
		self.Scheduler = Scheduler

		instance = self
	end
	return instance
end

return Roads.new()
